### ISSUE TITLE
which problem should I use to classify text?

### ISSUE THREAD

**Issue by yiqiang-zhao** [tensorflow/tensor2tensor#364]
Hi all, I’m trying to do a binary text classification using t2t, but I can’t find any problem already defined of this type.
And I want to use attention without LSTM or CNN, so I choose the transformer_encoder model.
I define a problem by myself, but it can’t work properly (test on training set, decoder only outputs one class label).
Can anyone give me some advice about my usage of t2t? 
Thank you.

Commands I used
```bash
t2t-datagen \
  --data_dir=$DATA_DIR \
  --tmp_dir=$TMP_DIR \
  --problem=text2_class_problem \
  --t2t_usr_dir=~/t2t_usr

t2t-trainer \
  --data_dir=$DATA_DIR \
  --problems=text2_class_problem \
  --model=transformer_encoder \
  --hparams_set=transformer_base \
  --t2t_usr_dir=~/t2t_usr \
  --output_dir=$TRAIN_DIR \
  --tmp_dir=$TMP_DIR \
  --train_steps=5000

t2t-decoder \
  --problems=text2_class_problem \
  --model=transformer_encoder \
  --hparams_set=transformer_base \
  --output_dir=$TRAIN_DIR \
  --decode_hparams="beam_size=1" \
  --t2t_usr_dir=~/t2t_usr \
  --decode_from_file=~/bw_data/1k_bw_train_data_r \
  --decode_to_file=~/bw_data/1k_bw_train_data_r_predict_res
```

my code:
```python
import os
import tensorflow as tf

from tensor2tensor.data_generators import problem
from tensor2tensor.utils import registry
from tensor2tensor.models import transformer
from tensor2tensor.data_generators import text_encoder
from tensor2tensor.data_generators import generator_utils

LOCATION_OF_DATA = '~/bw_data/'
_TRAIN_DATASETS = [LOCATION_OF_DATA + '1k_bw_train_data_r', LOCATION_OF_DATA + '1k_bw_train_label_id']
_TEST_DATASETS  = [LOCATION_OF_DATA + '1k_bw_train_data_r', LOCATION_OF_DATA + '1k_bw_train_label_id']

@registry.register_problem
class Text2ClassProblem(problem.Problem):
    m_vs_source = 268874

    @property
    def num_classes(self):
        return 2

    @property
    def class_labels(self):
        return ["ID_%d" % i for i in range(self.num_classes)]

    def feature_encoders(self, data_dir):
        source_vocab_size = Text2ClassProblem.m_vs_source
        source_vocab_filename = os.path.join(LOCATION_OF_DATA, "vocab.txt2tag.txt.%d" % source_vocab_size)
        source_token = text_encoder.TokenTextEncoder(source_vocab_filename)
        return {
            #"inputs": text_encoder.TextEncoder(),
            "inputs": source_token,
            "targets": text_encoder.ClassLabelEncoder(self.class_labels)
        }

    def generator(self, data_dir, tmp_dir, train):
        source_vocab_size = Text2ClassProblem.m_vs_source
        dataset = _TRAIN_DATASETS if train else _TEST_DATASETS
        fn_v_s = os.path.join(LOCATION_OF_DATA, "vocab.txt2tag.txt.%d" % source_vocab_size)
        source_vocab = text_encoder.TokenTextEncoder(fn_v_s)
        target_vocab = text_encoder.ClassLabelEncoder(self.class_labels)
        return text_class_token_generator(dataset[0], dataset[1], source_vocab, target_vocab)

    @property
    def input_space_id(self):
        return problem.SpaceID.ZH_TOK

    @property
    def target_space_id(self):
        return problem.SpaceID.IMAGE_LABEL

    @property
    def num_shards(self): 
        return 20 #100

    @property
    def num_dev_shards(self): 
        return 1

    def generate_data(self, data_dir, tmp_dir, task_id=-1): 
        train_paths = self.training_filepaths(data_dir, self.num_shards, shuffled=False)
        dev_paths = self.dev_filepaths(data_dir, self.num_dev_shards, shuffled=False)
        generator_utils.generate_dataset_and_shuffle(
            self.generator(data_dir, tmp_dir, True), train_paths,
            self.generator(data_dir, tmp_dir, False), dev_paths)

    def hparams(self, defaults, unused_model_hparams): 
        p = defaults

        p.input_modality = {"inputs": (registry.Modalities.SYMBOL, Text2ClassProblem.m_vs_source)}
        p.target_modality = (registry.Modalities.CLASS_LABEL, 2)
        #p.target_modality = registry.Modalities.CLASS_LABEL
        p.input_space_id = self.input_space_id
        p.target_space_id = self.target_space_id

def text_class_token_generator(source_path, target_path, source_token_vocab, target_token_vocab, eos=None):
    eos_list = [] if eos is None else [eos]
    with tf.gfile.GFile(source_path, mode="r") as source_file:
        with tf.gfile.GFile(target_path, mode="r") as target_file:
            source, target = source_file.readline(), target_file.readline()
            while source and target:
                source_ints = source_token_vocab.encode(source.strip()) + eos_list
                target_ints = [target_token_vocab.encode(target.strip())] + eos_list
                yield {"inputs": source_ints, "targets": target_ints}
                source, target = source_file.readline(), target_file.readline()
```

**Comment by yiqiang-zhao** [comment]
Finally, I find that imdb is a good example Problem of text classification.

**Comment by suryadst** [comment]
Hi. Did you implement tensor2tensor on the imdb text classification task? If so, can you share your implementation? Thank you!
